Add support for comments in .dockerignore

yongtang opened the pull request:
**\- What I did**

This fix tries to address the issue raised in #20083 where comment is not supported in `.dockerignore`.

**\- How I did it**

This fix updated the processing of `.dockerignore` so that any lines starting with `#` are ignored, which is similiar to the behavior of `.gitignore`.

Related documentation has been updated.

**\- How to verify it**

Additional tests have been added to cover the changes.

**\- Description for the changelog**

Add support for comments (lines starting with `#`) in .dockerignore.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #20083.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

yongtang commented:
cc @duglin @thaJeztah.

LK4D4 commented:
Probably people will expect inline comments to work as well, but I'm not sure.

vdemeester commented:
@LK4D4 yeah I guess, but git didn't support inline coment for a while, depending on the diffculty, I think having only lines comment is a good start.

thaJeztah commented on builder/dockerignore/dockerignore.go:
```diff
@@ -21,7 +21,8 @@ func ReadAll(reader io.ReadCloser) ([]string, error) {
 		pattern := strings.TrimSpace(scanner.Text())
-		if pattern == "" {
+		// Blank lines or lines starting with # are ignored.
+		if pattern == "" || strings.HasPrefix(pattern, "#") {
```
the `strings.HasPrefix(`.. should be before we trim spaces, otherwise we're not consistent with the description;

```
# this is a comment
            # this is now also considered a comment
```

duglin replied:
I'm ok with either. I always thought it was odd that Dockerfiles didn't support spaces before comments.

yongtang replied:
Thanks @thaJeztah @duglin I updated the pull request. Now only lines starting with `#` (at position 1) are considered as comments.

thaJeztah commented:
I'm fine with having just this as a start; Dockerfile's also only support comments if they start at position 1. Supporting inline comments doesn't look complicated if we accept that paths cannot contain `#`;

```go
pattern = strings.SplitN(pattern, "#", 2)[0]
```

@vdemeester @LK4D4 wdyt?

justincormack commented:
Supporting inline comments seems a bad thing for something as simple as this, let's just do the simplest thing and accept comments only if the first character is `#`.

thaJeztah commented:
Ok, let's go with that option for now (as mentioned, it's also consistent with the Dockerfile syntax). 

Should we address https://github.com/docker/docker/pull/23111#discussion_r65146798? I'm a bit in doubt as it's inconsistent with the way Dockerfiles are parsed, although the documentation mentions;

> A preprocessing step removes leading and trailing whitespace

So I'm a bit in doubt as to what the most sensible approach is

justincormack commented:
I think the preprocessing step should only take place on non comment lines, so the description should be first comments are removed then the current processing takes place.

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -6740,3 +6740,30 @@ func (s *DockerSuite) TestBuildLabelsOverride(c *check.C) {
+# because we have git ;-)
+# The above comment is from #20083
+./foo
+#dir1//foo
```
why the double / ?
Let's keep the test simple and just make it `#dir1/foo` to make sure its really ignored.
And if you do end up supporting spaces before "#" then please include that in the test too.
And you should probably add one where "#" is in the middle of the line - to make sure we *don't* read it as a comment.

yongtang replied:
Thanks @duglin. The pull request has been updated with simplified filename. Also two files starting with `#` (one `#1` and another `#2`) have been added to test `#` in the middle case.

justincormack commented on docs/reference/builder.md:
````diff
@@ -379,9 +379,13 @@ the working and the root directory.  For example, the patterns
 Here is an example `.dockerignore` file:
 
 ```
+    # comment
````
We should remove the leading spaces from the example, it is confusing, even though previously valid.

yongtang replied:
Thanks @justincormack the leading spaces have been removed.

yongtang commented:
Thanks @LK4D4 @vdemeester @duglin @thaJeztah @justincormack for the review. The pull request has been updated.

Now only lines starting with `#` (at position 1) are considered as comments and are ignored.

In addition, I updated the test by adding a couple of files with file name `#1` and `#2`. This is to test the case of `#` in the middle of the line (which is not considered as comment and is not ignored therefore).

Please let me know if there are any issues.

duglin commented on builder/dockerignore/dockerignore.go:
```diff
@@ -20,6 +20,10 @@ func ReadAll(reader io.ReadCloser) ([]string, error) {
 
 	for scanner.Scan() {
+		// Lines starting with # (comments) are ignored before processing
+		if strings.HasPrefix(scanner.Text(), "#") {
```
its a minor performance (stylistic) thing, but can you change this so that its:
```
pattern := scanner.Text()
if strings.HasPrefix(pattern, "#") { ... }
pattern = strings.TrimSpace(pattern)
```
So that we're not calling scanner.Text() more than once?

yongtang replied:
Thanks @duglin the pull request has been updated.

duglin commented on docs/reference/builder.md:
```diff
@@ -379,9 +379,13 @@ the working and the root directory.  For example, the patterns
 in the `foo` subdirectory of `PATH` or in the root of the git
 repository located at `URL`.  Neither excludes anything else.
 
+If a line in `.dockerignore` file starts with `#`, then this line is
```
... starts with `#` in column 1, then ...
(added "in column 1")

yongtang replied:
Thanks @duglin the docs has been updated.

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -6740,3 +6740,37 @@ func (s *DockerSuite) TestBuildLabelsOverride(c *check.C) {
+# The following is considered as comment as # is at the beginning
+#1
+# The following is not considered as comment as # is not at the beginning
+  #2
```
This inconsistency made me laugh.  Leading spaces are meaningful  when looking for comments, but not in filenames. IOW, if leading spaces aren't just meaningless noise that we skip, then technically the file name in this case should start with spaces too - which, btw, is totally legal.

justincormack replied:
Yes, it is a bit odd, but we have allowed leading spaces before filenames already, so we don't really want to change this behaviour, and while it does not provide a way to ignore files with spaces at the start, it does allow a workaround to ignore files with # at the start...!

duglin replied:
LOL I guess, but I think that's less likely than leading spaces for comments, which I know people dislike about Dockerfiles.  I just don't see the benefit of keeping this odd behavior. Eventually Dockerfile processing will be changed, but these .dockerignore files will probably live on longer after Dockerfiles are "done right".  I'd suggest we do it right for these.

yongtang replied:
Thanks @duglin @justincormack. I could see it is kind of confusing sometimes. Don't know what is the best way to cover those scenarios.

Maybe we could just always remove leading space, or we could introduce a escape character like `\`?

yongtang commented:
The failure in windowsTP5 seems to be real. I would guess it is related to the `(ls /tmp/#1)` and Windows may not recognize `#`.

I don't have a setup in windowsTP5 locally yet. But I will try to do some investigation on this issue.

```
15:29:05 FAIL: docker_cli_build_test.go:6745: DockerSuite.TestBuildDockerignoreComment
15:29:05 
15:29:05 docker_cli_build_test.go:6774:
15:29:05     c.Fatal(err)
15:29:05 ... Error: failed to build the image: Sending build context to Docker daemon 5.632 kB

15:29:05 Step 1 : FROM busybox
15:29:05  ---> 70c337fe05f8
15:29:05 Step 2 : ADD . /tmp/
15:29:05  ---> b47bf1e14ad7
15:29:05 Removing intermediate container bf8b6e903af4
15:29:05 Step 3 : RUN sh -c "(! ls /tmp/foo) && (! ls /tmp/foo2) && (ls /tmp/dir1/foo) && (ls /tmp/#1) && (! ls /tmp/#2)"
15:29:05  ---> Running in 0277780112c5
15:29:05 ls: /tmp/foo2: No such file or directory
15:29:05 The command 'cmd /S /C sh -c "(! ls /tmp/foo) && (! ls /tmp/foo2) && (ls /tmp/dir1/foo) && (ls /tmp/#1) && (! ls /tmp/#2)"' returned a non-zero code: 255
15:29:05 
15:29:05 
```

yongtang committed:
Add support for comment in .dockerignore

This fix tries to address the issue raised in #20083 where
comment is not supported in `.dockerignore`.

This fix updated the processing of `.dockerignore` so that any
lines starting with `#` are ignored, which is similiar to the
behavior of `.gitignore`.

Related documentation has been updated.

Additional tests have been added to cover the changes.

This fix fixes #20083.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
The issue in TP5 should have been fixed.

duglin commented:
Where did janky go?

thaJeztah commented:
Took a day off, possibly Leeroy (@GordonTheTurtle) didn't get a call. I'll rebuild

duglin commented:
LGTM if janky reappears and is ok.

yongtang commented:
Thanks @thaJeztah @duglin.

I also noticed that documentation always fails, not just for this PR but possibly for any PR related to documentation change:

https://github.com/docker/docker/pull/23192
https://github.com/docker/docker/pull/23195
https://github.com/docker/docker/pull/23107

And the documentation error always points to 
```
ERROR: 2016/06/02 05:39:15 content_renderer.go:56: LinkResolver: No page found for "/docker-trusted-registry/overview.md" on page "registry/index.md".
```

I am not familiar with how Docker's Jenkins build for documentation works so not sure how to fix it.

thaJeztah commented:
@yongtang yes, It looks like the CI settings changed and now fail instead of "warn" for some errors. That link error looks to be in another repository, so should probably be fixed in the "distribution" repository.

yongtang commented:
Thanks @thaJeztah @duglin. The pull request has been rebased and the documentation issue seems to have been fixed. Please let me know if there are any other issues.

vdemeester commented:
LGTM 🐮

vdemeester commented:
Moving to doc review /cc @thaJeztah

thaJeztah commented:
docs LGTM

gccgo timed out

vdemeester commented:
docs LGTM it is, merging :angel:
